rust/parquet/src/column/reader.rs

```diff
@@ -190,15 +190,12 @@ impl<T: DataType> ColumnReaderImpl<T> {
                     (self.num_buffered_values - self.num_decoded_values) as usize,
                 );
 
-                // Adjust batch size by taking into account how much space is left in
-                // values slice or levels slices (if available)
-                adjusted_size = min(adjusted_size, values.len() - values_read);
-                if let Some(ref levels) = def_levels {
-                    adjusted_size = min(adjusted_size, levels.len() - levels_read);
-                }
-                if let Some(ref levels) = rep_levels {
-                    adjusted_size = min(adjusted_size, levels.len() - levels_read);
-                }
+                // Adjust batch size by taking into account how much data there
+                // to read. As batch_size is also smaller than value and level
+                // slices (if available), this ensures that available space is not
+                // exceeded.
+                adjusted_size = min(adjusted_size, batch_size - values_read);
```

Comment:
I don't fully get this. In which case this code will yield different result than previous one. Could you give an example?

Reply:
As stated in my PR comment, the `read_batch` function can receive any combination of `batch_size`, `def_levels.len()`, `rep_levels.len()` and `values.len()`. If the `batch_size` is the limiting factor your `iter_batch_size` might end up larger than the `batch_size`. This happened to me when called from `record_reader.rs` on a parquet file with relatively small row groups (500k rows). I did not manage to reproduce the phenomenon with a mock data file though ... :-/

Reply:
Yes. I just struggled to come up with a concrete example for this. :)

Reply:
is it clearer now or do you need me to find a better example ?

Reply:
Could this be related to https://issues.apache.org/jira/browse/ARROW-8737 ?

Reply:
@sunchao Any objection to us getting this one merged?

Reply:
this very much looks to be the reason for your issue yes! Can you isolate the context in which your error occurred into a test ? I didn't find a way to do it with my data...